### DISCUSSION TITLE
How to amend two levels by also introduce local properties at the first level

### DISCUSSION THREAD

**Discussion by StefMa** [apple/pkl#736] [Q&A]
Wow, sorry for the title 🙈 

Given I have a template like this:
```pkl
module Jobber

jobs: Listing<Job>
```

Now I have two files that `amends` it:
```pkl
amends "modules/Jobber.pkl"

jobs {
 job1 {}
 job2 {}
 job3 {}
 job4 {}
}
```

```pkl
amends "modules/Jobber.pkl"

jobs {
 job1 {}
 job2 {}
 job3 {
  // Does something different than the other child
 }
 job4 {}
}
```

So basically, file b is identical to file a, except of a single job.

I thought I can extract a `BaseJobber` that amends `Jobber` and have somekind of `customJobs: Listing<Job>`.
My `BaseJobber` could look like that:
```pkl
amends "modules/Jobber.pkl"

customJobs: Listing<Job>

jobs {
 job1 {}
 job2 {}
 ...customJobs
 job4 {}
}
```

And the childs can amends `BaseJobber` and only override `customJobs` with their custom jobs.

But this doesn't work. I can't introduce new properties when I `amend` a module.

In my current solution I use `import Jobber` in my `BaseJobber` and define defaults.
Plus a function like 
```pkl
function jobs(injectedJob: Listing<Job>): Listing<Job> = new {
 job1 {}
 job2 {}
 ...customJobs
 job4 {}
}
```

With that the child can call `amends` the `Jobber` and `import` the `BaseJobber` and basically delegating everything to it.
This works, but doesn't look correct 🫠 
What is the pkl way to go to handle this case? 🤔 
Thanks in advance!

**Comment by bioball** [comment]
There's multiple ways to skin the cat!

One pattern is to define _another_ module that sets its output to the underlying module's output. It can then expose `customJobs` like you were thinking:

```pkl
// Jobber.pkl

module Jobber

jobs: Listing<Job>

class Job { name: String }
```

```pkl
// fancyJobber.pkl

import "Jobber.pkl"

customJobs: Listing<Jobber.Job>

fixed jobber: Jobber = new {
  jobs {
    new { name = "job 1" }
    new { name = "job 2" }
    ...customJobs
    new { name = "job 3" }
    new { name = "job 4" }
  }
}

output = jobber.output
```

```pkl
// myFancyJobber.pkl

amends "fancyJobber.pkl"

customJobs {
  new { name = "my custom job" }
}
```

**Reply by StefMa** [reply]
Thaks for the answer @bioball!
My example was a bit too simplified.
My `Jobber` has multiple properties.

If we introduce a `name: String` there, the whole thing doesn't work anymore.
Or at least I don't know how to.

I tried the following, but this don't work.
Of course, the output of `fancyJobber` is defined as a `Listing`.

```pkl
// Jobber.pkl

module Jobber

jobs: Listing<Job>

class Job { name: String }

name: String
```

```pkl
// fancyJobber.pkl

import "Jobber.pkl"

customName: String
fixed name: String = "Fancy Jobber"

customJobs: Listing<Jobber.Job>
fixed jobber: Jobber = new {
    jobs {
        new { name = "job 1" }
        new { name = "job 2" }
        ...customJobs
        new { name = "job 3" }
        new { name = "job 4" }
    }
}

output = jobber.output
```

```pkl
// myFancyJobber.pkl

amends "fancyJobber.pkl"

customName = "My Fancy Jobber"

customJobs {
    new { name = "my custom job" }
}
```

When running eval on the `myFancyJobber.pkl` 
```
pkl eval somewhere/modules/myFancyJobber.pkl
```

```
–– Pkl Error ––
Tried to read property `name` but its value is undefined.

7 | name: String
    ^^^^
at Jobber#name (file:///somewhere/modules/Jobber.pkl, line 7)

The above error occurred when rendering path `name`.

106 | text = renderer.renderDocument(value)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
at pkl.base#Module.output.text (https://github.com/apple/pkl/blob/0.26.3/stdlib/base.pkl#L106)
```
